emcache/cluster.py

```diff
@@ -221,6 +273,93 @@ def node(self, memcached_host_address: MemcachedHostAddress) -> Node:
+                    await asyncio.sleep(AUTODISCOVERY_POLL_INTERVAL_FAILURE)
+            except asyncio.CancelledError:
+                # if a cancellation is received we stop monitoring for cluster changes
+                break
+
+            except Exception:
+                logger.exception("Autodiscover raised an exception, continuing processing")
+
+        logger.debug("Autodiscovery task stopped")
+
+    async def _get_autodiscovered_nodes(self) -> tuple[bool, int, list[tuple[str, str, int]]]:
+        if self._closed:
+            raise RuntimeError("Emcache client is closed")
+
+        node = random.choice(self._discovery_nodes if self._discovery_nodes else self.nodes)
+        async with OpTimeout(self._autodiscovery_timeout, self._loop):
+            async with node.connection() as connection:
+                return await connection.autodiscovery()
+
+    async def autodiscover(self) -> bool:
+        try:
+            autodiscovery, version, nodes = await self._get_autodiscovered_nodes()
+        except asyncio.TimeoutError as e:
+            logger.error("Got timeout when checking cluster configuration: %r", e)
+            return False
+
+        if not autodiscovery:
+            logger.warning("autodiscovery request failed, does the server support 'config get cluster' command?")
+            return False
+
+        if version == self._autodiscover_config_version:
+            return False
+
+        if not nodes:
+            logger.error("Node list received from autodiscovery is empty!")
+            return False
+
+        # Updating node list with preservation of their status
+        new_nodes_set = {MemcachedHostAddress(host, port) for host, ip, port in nodes}
+        old_nodes_set = {node.memcached_host_address for node in self.nodes}
+        protected = {node.memcached_host_address for node in self._discovery_nodes}
+
+        # Add brand-new nodes to the list as healthy
+        for node in new_nodes_set - old_nodes_set:
+            self._healthy_nodes.append(Node(node, *self._new_node_options))
+
+        closable: list[Node] = []
+
+        # Remove unhealthy nodes that are no longer being used
+        new_nodes: list[Node] = []
+        for node in self._unhealthy_nodes:
+            if node.memcached_host_address in new_nodes_set:
+                new_nodes.append(node)
+            elif node.memcached_host_address not in protected:
+                closable.append(node)
+        self._unhealthy_nodes = new_nodes
+
+        # Remove healthy nodes that are no longer being used
+        new_nodes: list[Node] = []
+        for node in self._healthy_nodes:
+            if node.memcached_host_address in new_nodes_set:
+                new_nodes.append(node)
+            elif node.memcached_host_address not in protected:
+                closable.append(node)
+        self._healthy_nodes = new_nodes
+
+        # Close unused nodes
+        for node in closable:
+            await node.close()
```

Comment:
Would make sense to move this to the very end of function, until `self._build_rdz_nodes()` is not executed commands issued might use invalid nodes.

Reply:
done